### PR TITLE
initial test

### LINKED ISSUE #394: Component unit tests: ProtectedRoute
We need unit tests for the [ProtectedRoute](https://github.com/beyondhb1079/s4us/blob/main/src/components/ProtectedRoute.jsx). Consider any edge cases. See the [Testing Library docs](https://testing-library.com/docs/) and existing tests for reference.

### PR DESCRIPTION
**Pull request by edwingl10**
## Motivation and Context

Fixes #394 

## How Has This Been Tested?

On laptop using Firefox. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] User visible change (users will notice UI or functional changes)

## Previewing Changes

1. Go to https://s4us-pr-NUMBER.onrender.com.
2.

## Checklist:

- [x] I have added tests to cover my changes.

### CODE REVIEW

File: src/components/ProtectedRoute.test.jsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import { MemoryRouter, Route } from 'react-router-dom';
+import { clearFirestoreData, initializeTestApp } from '../lib/testing';
+import ScholarshipForm from '../pages/ScholarshipForm';
+
+window.MutationObserver = require('mutation-observer');
+
+function renderAtRoute(route) {
+  return render(
+    <MemoryRouter initialEntries={[route]}>
+      <Route path="/scholarships/new" component={ScholarshipForm} />
```
Comment (josueetcom) [inline src/components/ProtectedRoute.test.jsx:12]: A few things: 
1. This should be a **ProtectedRoute**
2. The `path` is technically arbitrary but I'd expect it to be `route`?
3. The `component` is also arbitrary but instead of the **ScholarshipForm** (which has its own requirements) I suggest having this method accept a `component` which in the tests can something as simple as `<h1>Protected header</h1>`

The way this is currently written this looks more like what a test might look like for **AddScholarship** (as that's what this page was renamed to).

File: src/components/ProtectedRoute.test.jsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    </MemoryRouter>
+  );
+}
+
+const app = initializeTestApp({
+  projectId: 'protected-route-test',
+  apiKey: 'fake-api-key',
+});
+
+beforeAll(() => clearFirestoreData(app.options));
```
Comment (josueetcom) [inline src/components/ProtectedRoute.test.jsx:22]: Firestore is irrelevant here.

File: src/components/ProtectedRoute.test.jsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+afterAll(() => app.delete());
+
+test('athenticated user can accesss page', () => {
```
Comment (josueetcom) [inline src/components/ProtectedRoute.test.jsx:25]: Does this test work?

Technically there are three test cases:
1. Initially the **CircularProgress** is displayed when `user` is undefined.
2. "You must log in first to access this page" when the `user` is not signed in.
3. The passed in component is displayed when `user` is signed in.

I'm not sure exactly how to mock the user but something you can manually set `firebase.auth().currentUser`.

Maybe you can even test the `onAuthStateChanged` thing using [firebase.auth().updateCurrentUser()](https://firebase.google.com/docs/reference/js/auth.auth.md#authupdatecurrentuser) and collapse the first test case with the other two.

File: src/components/ProtectedRoute.test.jsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  expect(screen.getByLabelText('Deadline *')).toBeInTheDocument();
+  expect(screen.getByLabelText('Description *')).toBeInTheDocument();
+  expect(screen.getByLabelText('Website *')).toBeInTheDocument();
+
+  expect(screen.getByText('Amount Type *')).toBeInTheDocument();
+  expect(screen.getByText('Unknown')).toBeInTheDocument();
+  expect(screen.getByText('Fixed Amount:')).toBeInTheDocument();
+  // expect(screen.getByText('Range:')).toBeInTheDocument();
+  expect(screen.getByText('Full Tuition')).toBeInTheDocument();
+  expect(screen.getByRole('button', { name: /SUBMIT/i }));
```
Comment (josueetcom) [inline src/components/ProtectedRoute.test.jsx:39]: If you use a simple component you only need one statement.

**Review by josueetcom — CHANGES_REQUESTED** [review]

**Comment by josueetcom** [comment]
I think the fake auth testing behavior in #698 should help with writing these test cases!

File: src/components/ProtectedRoute.test.jsx (hunk 3 above)
Comment (josueetcom) [reply src/components/ProtectedRoute.test.jsx:26]: *authenticated

File: src/components/ProtectedRoute.test.jsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+afterAll(() => app.delete());
+
+test('athenticated user can accesss page', () => {
+  renderAtRoute('/authenticated');
+
+  expect(screen.getByText('Protected Header')).toBeInTheDocument();
+});
+
+test("unathenticated user can't access page", () => {
+  app.auth().signOut();
```
[the remark this answers is not in the document]
Comment (josueetcom) [reply src/components/ProtectedRoute.test.jsx:33]: Hmm it might be the fact **ProtectedRoute** uses `firebase.auth()`. In #698 I changed `Scholarships.ts` to use `firebase.app().auth()` instead of `firebase.auth()`:
https://github.com/beyondhb1079/s4us/blob/eb8ba2ffe3a1b6de8f32c9135a39c1164c024a3b/src/models/Scholarships.ts#L12

I think I had to do that because I mocked the test app's auth function:
https://github.com/beyondhb1079/s4us/blob/eb8ba2ffe3a1b6de8f32c9135a39c1164c024a3b/src/lib/testing.ts#L45

Maybe adding a `firebase.auth = app.auth` statement before `return app` in the `initializeTestApp` method would fix this? 🤔

File: src/components/ProtectedRoute.test.jsx [hunk 6]
```diff
@@ -1,40 +1,38 @@
...
+
+  renderAtRoute('/unauthenticated');
+
+  expect(screen.getByText('Protected Header')).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (edwingl10) [reply src/components/ProtectedRoute.test.jsx:37]: yes but since it was failing, I just left it like this for now.

File: src/components/ProtectedRoute.test.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (edwingl10) [reply src/components/ProtectedRoute.test.jsx:33]: hmm I tried it but it throws an error
Screen Shot 2021-10-31 at 11 09 37 AM
